text/0000-async-stream.md

```diff
@@ -0,0 +1,754 @@
+
+In summary, an async stream:
+* has a pinned receiver
+* that takes `cx` context
+* and returns a `Poll`
+
+## Initial impls
+
+There are a number of simple "bridge" impls that are also provided:
+
+```rust
+impl<S> Stream for Box<S>
+where
+    S: Stream + Unpin + ?Sized,
+{
+    type Item = <S as Stream>::Item
+}
+
+impl<S> Stream for &mut S
+where
+    S: Stream + Unpin + ?Sized,
+{
+    type Item = <S as Stream>::Item;
+}
+
+impl<S, T> Stream for Pin<P>
+where
+    P: DerefMut<Target=T> + Unpin,
+    T::Target: Stream,
+{
+    type Item = <T as Stream>::Item;
+}
+
+impl<S> Stream for AssertUnwindSafe<S>
+where
+    S: Stream, 
+{
+    type Item = <S as Stream>::Item;
+}
+```
+
+## Next method/struct
+
+We should also implement a next method, similar to [the implementation in the futures-util crate](https://docs.rs/futures-util/0.3.5/src/futures_util/stream/stream/next.rs.html#10-12).
+
+In general, we have purposefully kept the core trait definition minimal. 
+There are a number of useful extension methods that are available, for example, 
+in the futures-stream crate, but we have not included them because they involve 
+closure arguments, and we have not yet finalized the design of async closures.
+
+However, the core methods alone are extremely unergonomic. You can't even iterate 
+over the items coming out of the stream. Therefore, we include a few minimal 
+convenience methods that are not dependent on any unstable features, such as `next`.
+
+```rust
+/// A future that advances the stream and returns the next value.
+///
+/// This `struct` is created by the [`next`] method on [`Stream`]. See its
+/// documentation for more.
+///
+/// [`next`]: trait.Stream.html#method.next
+/// [`Stream`]: trait.Stream.html
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct Next<'a, S: ?Sized> {
+    stream: &'a mut S,
+}
+
+impl<St: ?Sized + Unpin> Unpin for Next<'_, St> {}
+
+impl<'a, St: ?Sized + Stream + Unpin> Next<'a, St> {
+    pub(super) fn new(stream: &'a mut St) -> Self {
+        Next { stream }
+    }
+}
+
+impl<St: ?Sized + Stream + Unpin> Future for Next<'_, St> {
+    type Output = Option<St::Item>;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Self::Output> {
+        Pin::new(&mut *self.stream).poll_next(cx)
+    }
+}
+```
+
+This would allow a user to await on a future:
+
+```rust
+while let Some(v) = stream.next().await {
+
+}
+```
+
+We could also consider adding a try_next? function, allowing
+a user to write:
+
+```rust
+while let Some(x) = s.try_next().await?
+```
+
+One thing to note, if a user is using an older version of `futures-util`,
+they would experience ambiguity when trying to use the `next` method that
+is added to the standard library (and redirected to from `futures-core`).
```

Comment:
Given that updating `futures-core` to use `core::?::Stream` will be a major version bump this should not be an issue in practice. (Unless a library implements both `futures-core0.3::Stream` and `std::Stream` on a type).

Reply:
Understood, while it should not be an issue in most cases, I wanted to include it just in case we see problems with edge cases in the future.

Reply:
There's some discussion in <https://github.com/rust-lang/futures-rs/issues/2207#issuecomment-689573269>. It is definitely a "technically" breaking change, and my instinct would be that it _will_ cause confusing breakage in practice when users don't `cargo update` to a version of `futures-core` that uses a re-export.

Reply:
Can we link to the above discussion in the RFC?